src/core/ee/cop1.cpp

```diff
@@ -197,215 +218,224 @@ void Cop1::div_s(int dest, int reg1, int reg2)
             control.d = true;
             control.sd = true;
         }
-        gpr[dest].u = ((numerator ^ denominator) & 0x80000000) | 0x7F7FFFFF;
+        uint32_t result = ((numerator ^ denominator) & 0x80000000) | 0x7F7FFFFF;
+        memcpy(&gpr[dest], &result, sizeof(float));
     }
     else
-        gpr[dest].f = convert(numerator) / convert(denominator);
+        gpr[dest] = convert(gpr[reg1]) / convert(gpr[reg2]);
 
-    check_overflow(gpr[dest].u, false);
-    check_underflow(gpr[dest].u, false);
+    check_overflow(gpr[dest], false);
+    check_underflow(gpr[dest], false);
 
-    printf("[FPU] div.s: %f(%d) / %f(%d) = %f(%d)\n", convert(numerator), reg1, convert(denominator), reg2,
-           gpr[dest].f, dest);
+    printf("[FPU] div.s: %f(%d) / %f(%d) = %f(%d)\n", convert(numerator), reg1, convert(denominator), reg2, gpr[dest], dest);
 }
 
-void Cop1::sqrt_s(int dest, int source)
+void Cop1::sqrt_s(int dest, int source) noexcept
 {
-    if ((gpr[source].u & 0x7F800000) == 0)
-        gpr[dest].u = gpr[source].u & 0x80000000;
+    uint32_t usource;
+    uint32_t udest;
+    memcpy(&usource, &gpr[source], sizeof(float));
+    memcpy(&udest, &gpr[dest], sizeof(float));
+    if ((usource & 0x7F800000) == 0)
+    {
+        udest = usource & 0x80000000;
+        memcpy(&gpr[dest], &udest, sizeof(float));
+    }
     else
     {
-        if (gpr[source].u & 0x80000000)
+        if (usource & 0x80000000)
         {
-            printf("[FPU] Negative Sqrt Fs = %x\n", gpr[source].u);
+            printf("[FPU] Negative Sqrt Fs = %x\n", usource);
             control.i = true;
             control.si = true;
         }
 
-        gpr[dest].f = sqrt(fabs(convert(gpr[source].u)));
+        gpr[dest] = sqrtf(fabsf(convert(gpr[source])));
     }
 
     control.d = false;
-    printf("[FPU] sqrt.s: %f(%d) = %f(%d)\n", gpr[source].f, source, gpr[dest].f, dest);
+    printf("[FPU] sqrt.s: %f(%d) = %f(%d)\n", gpr[source], source, gpr[dest], dest);
 }
 
-void Cop1::abs_s(int dest, int source)
+void Cop1::abs_s(int dest, int source) noexcept
 {
-    gpr[dest].u = gpr[source].u & 0x7FFFFFFF;
+    gpr[dest] = fabsf(gpr[source]);
 
     control.u = false;
     control.o = false;
 
-    printf("[FPU] abs.s: %f = -%f\n", gpr[source].f, gpr[dest].f);
+    printf("[FPU] abs.s: %f = -%f\n", gpr[source], gpr[dest]);
 }
 
-void Cop1::mov_s(int dest, int source)
+void Cop1::mov_s(int dest, int source) noexcept
 {
-    gpr[dest].u = gpr[source].u;
+    gpr[dest] = gpr[source];
     printf("[FPU] mov.s: %d = %d", source, dest);
 }
 
-void Cop1::neg_s(int dest, int source)
+void Cop1::neg_s(int dest, int source) noexcept
 {
-    gpr[dest].u = gpr[source].u ^ 0x80000000;
+    gpr[dest] = -gpr[source];
```

Comment:
NEG only flips the sign bit, not the whole value, not sure how this would behave, and what if the original value is already negative? surely doing - will keep it the same, which is not a negation